Add xarm6 robot with different grippers

dwaitbhatt opened the pull request:
Add xarm6 with left and right allegro grippers, xarm gripper, and a test version without a gripper.
This resolves the request in #635

StoneT2000 commented:
Thank you for the contribution @dwaitbhatt, I'll try and get around reviewing this later today or next monday!

I think only thing I need to verify really is to check if the new files aren't too large (if they are, we just make them downloadable instead of being on git), and if the simulation speed + collision meshes are reasonable.

dwaitbhatt commented:
> Thank you for the contribution @dwaitbhatt, I'll try and get around reviewing this later today or next monday!
> 
> I think only thing I need to verify really is to check if the new files aren't too large (if they are, we just make them downloadable instead of being on git), and if the simulation speed + collision meshes are reasonable.

Sounds good @StoneT2000! Some notes from my own testing:

- xarm6 with the default xarm gripper (`xarm6_xarmgripper`) is somehow not stable in simulation - when running with random actions, the gripper joints tend to go out of limits and it eventually leads all robot joints going blowing up and going to nan - similar to #633, although `xarm7_ability` becomes unstable much faster, within 5 steps. I suspect the issue there is also with the ability hand, not the arm. In my case, I am using the gripper URDF from the manufacturer's [own repo](https://github.com/xArm-Developer/uf-gym/blob/master/urdf/xarm/xarm6_with_gripper.urdf).

- xarm6 with both allegro hands is stable, but for some reason the simulation is faster with CPU (60fps) than GPU (~20fps), atleast when tested with `demo_robot_script` and random actions.

StoneT2000 commented:
Thanks for the details. 

Another question, for the actual joint controller of the xarm6 gripper, does the end effector have 1 joint (like a mimic joint) or are there multiple controlling


I recall that to correctly model that gripper we need to model the mimic joints correctly and correctly model that in sim

dwaitbhatt commented:
> Another question, for the actual joint controller of the xarm6 gripper, does the end effector have 1 joint (like a mimic joint) or are there multiple controlling
> 
> I recall that to correctly model that gripper we need to model the mimic joints correctly and correctly model that in sim.

It does have a single mimic joint to control the gripper, however the implementation is from the manufacturer's [repo](https://github.com/xArm-Developer/uf-gym/blob/master/urdf/xarm/xarm6_with_gripper.urdf) and not my own. Need to look into this more carefully.
 

> Finally where are the urdf / assets from? Usually need to add a readme somewhere just for record keeping about where it's from and what modifications are made if any to the original URDF.

Here are the sources for the assets and the modifications on top:
1) `xarm6_xarmgripper`: 
	- Source: xArm manufacturer's [repo](https://github.com/xArm-Developer/uf-gym/blob/master/urdf/xarm/xarm6_with_gripper.urdf)
	- Modifications:
		- Removed gazebo/ros related tags
		- Added `safety_controller` for safe operation beyond joint limits
		- Added a material (specular/ambient/diffuse color) based on provided MTL files

2) `xarm6_nogripper`: 
	- Source: Same as `xarm6_xarmgripper`, without gripper links and joints

3) `xarm6_allegro_left`: 
	- Source: urdf from a previous [issue](https://github.com/haosulab/ManiSkill/issues/482#issuecomment-2272987252), which was [confirmed](https://github.com/haosulab/ManiSkill/issues/482#issuecomment-2276533522) to work
	- Modifications:
		- Updated collision geometries to match `xarm6_xarmgripper`
		- Updated orientation of gripper so that the gripper wrist is connected to final link of the arm

4) `xarm6_allegro_right`: 
	- Source: Arm description from `xarm6_nogripper` with right hand allegro gripper description from  Maniskill [assets](https://github.com/haosulab/ManiSkill/blob/main/mani_skill/assets/robots/allegro/allegro_hand_right_glb.urdf)

StoneT2000 commented:
Thanks for the info. RE the gripper modelling, I have the correct code for how to model that which was uploaded in response to #477, its the floating version of the xarm gripper.

StoneT2000 commented on mani_skill/assets/robots/xarm6/xarm6_allegro_left_2023.urdf:
any reason why this has _2023 attached to the name?

dwaitbhatt replied:
Not really, it was just named that way in the code shared in a previous [issue](https://github.com/haosulab/ManiSkill/issues/482#issuecomment-2272987252). Removed the _2023 for consistent naming.

StoneT2000 commented on mani_skill/assets/robots/xarm6/xarm6_nogripper.urdf:
```diff
@@ -0,0 +1,250 @@
+<?xml version="1.0" encoding="utf-8"?>
```
all the URDFs should have some .srdf files which let you define custom collisions between links to ignore. this is likely why the sim speed isn't as fast as it could be as well. Definitely should ignore a lot of these that show up in the viewer if you open the contacts tab, some of them never occur but take up processing time. We by default already ignore adjacent links.

![image](https://github.com/user-attachments/assets/3ee63b8f-cbed-465c-927d-80f4bd7ac238)

See the fetch.srdf file in this repo for an example.

dwaitbhatt replied:
Fixed! Added SRDFs for all xarm6 variants. I have added more details about these in the comment [below](https://github.com/haosulab/ManiSkill/pull/636#issuecomment-2434495095).

StoneT2000 commented on mani_skill/agents/robots/xarm6/xarm6_xarmgripper.py:
```diff
@@ -0,0 +1,316 @@
+
+
+    @property
+    def _sensor_configs(self):
```
normally we like to have a with and without mounted camera variant of the robot.

A simple solution is to create a new class below this one called xarm6_xarmgripper_wristcam or something and have it inherit the original class but override the sensor configs.

dwaitbhatt replied:
Fixed! Separated xarm6 variants with and without wristcam.

akuramshin commented:
@StoneT2000 @dwaitbhatt Hello! I loaded the embodiment `xarm6_xarmgripper` into the StackCube-v1 environment using the `demo_manual_control.py` script (using `pd_ee_delta_pose` as the default controller). 

Proceeding to press the f/g keys to open/close the gripper causes the robot's ee pose to change despite only changing the gripper position.
[out2.webm](https://github.com/user-attachments/assets/473267de-d8a3-40a8-b869-7bc740d33dea)

The reason I am here is because I have been experiencing something similar with my agent setup of the robotiq 2F-85 gripper on the panda arm:
[out3.webm](https://github.com/user-attachments/assets/eebcd3b7-4b9e-459a-b3bd-1610df2a4013)

In both videos I am just periodically pressing either the f or g key.

Please let me know if this belongs in its own post. I just thought it was relevant since this PR also has the unexpected behaviour.

StoneT2000 commented:
Hi @akuramshin , how did you implement the robotiq one? Did you set the drives correctly / are you using our URDFs in this repo?

akuramshin commented:
I followed your thread here: https://github.com/haosulab/ManiSkill/issues/477#issuecomment-2299807384

Basically replacing the original panda hand with the urdf of the robotiq gripper you provided. For the controller I copy pasted the controller implemented for the `FloatingRobotiq2F85Gripper` agent

StoneT2000 commented:
Ok for your issue @akuramshin can you create a different issue on this repo? This PR will just be for the xarm gripper which is different. I will follow up there.

StoneT2000 commented:
Sorry I am a little bit behind on the PRs here, I will try to take a deeper look into this later this week.

dwaitbhatt commented:
> Sorry I am a little bit behind on the PRs here, I will try to take a deeper look into this later this week.

Hello @StoneT2000! Did you get a chance to review this PR?

StoneT2000 commented:
@dwaitbhatt did you change the gripper? I just checked and its not the old xarm gripper. The robot iq one version seems to work fine although i just ran pick cube with the xarm 6 robotiq gripper and it is not working. Currently debugging


https://github.com/user-attachments/assets/400c15ec-4d6a-4384-99c9-89cc72d6faa3

It is a older version of ManiSkill however, I will try to merge the main branch, which includes some optimizations and also a much better PPO.

Re sim speed there are a lot of extra contacts that can't ever happen, am removing them now with a .srdf file.

StoneT2000 commented:
oh i find the issue with why pick cube doesn't train. It's because the "eef" link defined pose is a bit too far

![image](https://github.com/user-attachments/assets/d88417f3-ccb0-476f-85e8-2c2405fc57e2)
visualized here. Hence the robot learns to go close but not actually go and grasp the cube, the first part of the reward fn is to minimize the distance between the tcp (the eef link here) and the cube.

StoneT2000 commented:
Oh I find the reason why pick cube wasn't solvable. 
- The is_grasping function is slightly wrong. The fingers being checked are not the pads (their the link behind the pads) and the link pose direction is reversed for some reason. 
- Another thing is the is_static condition is a bit different than panda in that you need to ignore all of the gripper links, passive and controlled ones. Otherwise in PickCube robot doesn't learn to sit still.
- The implemented xarm moves slower relative to the panda arm. This is really not a bug or issue, but it means the default table top environments likely have max episode steps set too low (they are optimized for the default robot used to solve the task). 
- The slower speed of the xarm (which can be fixed if you bump up the max delta joint positions in the controller) causes RL to be a bit harder although not impossible. 
- 
```
python ppo_fast.py --env_id="PickCube-v1" --seed=1    --num_envs=4096 --num-steps=16 --update_epochs=8 --num_minibatches=32     --total_timesteps=50_000_000     --num_eval_envs=16     --save-model --cudagraph
s --exp-name="ppo-PickCube-v1-state-xarm6-1-walltime_efficient" --num_eval_steps=100
```

https://github.com/user-attachments/assets/c47c5e43-ee11-4fb2-a171-618ff24ad905

dwaitbhatt commented:
> @dwaitbhatt did you change the gripper? I just checked and its not the old xarm gripper. The robot iq one version seems to work fine although i just ran pick cube with the xarm 6 robotiq gripper and it is not working. Currently debugging
> Re sim speed there are a lot of extra contacts that can't ever happen, am removing them now with a .srdf file.


@StoneT2000 Yes, I had changed the gripper because of issues with drive creation for xarm_gripper. I had also written SRDFs for xarm6_robotiq and xarm6_allegro. Details in https://github.com/haosulab/ManiSkill/pull/636#issuecomment-2434495095.

Regarding the rest, thank you for the fixes! Glad to see xarm6 finally learning in Maniskill. Let me know if you need any help with the final xarm6 PR.

StoneT2000 commented:
closed in favor of #820
